servicenow/stackstate_checks/servicenow/servicenow.py

```diff
@@ -117,123 +167,209 @@ def filter_empty_metadata(self, data):
+        for relation in collected_relations.get('result'):
             parent_sys_id = relation['parent']['value']
             child_sys_id = relation['child']['value']
             type_sys_id = relation['type']['value']
 
             relation_type = relation_types[type_sys_id]
             data = self.filter_empty_metadata(relation)
-            data.update({"tags": instance_config.instance_tags})
+            data.update({"tags": instance_info.instance_tags})
 
             self.relation(parent_sys_id, child_sys_id, relation_type, data)
 
+    def _collect_change_requests(self, instance_info):
+        # TODO: add limit for max number of events that we'll process
+        auth = (instance_info.user, instance_info.password)
+        url = instance_info.url + '/api/now/table/change_request'
+        params = '?sysparm_display_value=all&sysparm_exclude_reference_link=true'
+        params += '&sysparm_limit={}'.format(instance_info.change_request_process_limit)
+        if instance_info.state.latest_sys_updated_on:
+            reformatted_date = ','.join(
+                "'{}'".format(i) for i in str(instance_info.state.latest_sys_updated_on).split(' ')
+            )
+            quoted_date = quote(reformatted_date)
+            params += '&sysparm_query=sys_updated_on>javascript%3Ags.dateGenerate({})'.format(quoted_date)
+        url += params
+        self.log.info('url for getting CRs: %s', url)
+        return self._get_json(url, instance_info.timeout, auth)
+
+    def _sanitize_response(self, cr_list):
+        sanitized_crs = []
+        for cr in cr_list:
+            sanitized_crs.append(self.filter_empty_metadata(cr))
+        return sanitized_crs
+
+    def _process_change_requests(self, instance_info):
+        response = self._collect_change_requests(instance_info)
+        sanitized_result = self._sanitize_response(response['result'])
+        self.log.info('CRs results: %d', len(sanitized_result))
+        for cr in sanitized_result:
+            try:
+                change_request = ChangeRequest(cr, strict=False)
+                change_request.validate()
+            except DataError as e:
+                self.log.warning('%s - DataError: %s. This CR is skipped.', cr['number']['value'], e)
+                continue
+            if change_request.cmdb_ci:
+                if change_request.sys_updated_on > instance_info.state.latest_sys_updated_on:
+                    instance_info.state.latest_sys_updated_on = change_request.sys_updated_on
+                old_state = instance_info.state.change_requests.get(change_request.number)
+                if old_state is None or old_state != change_request.state:
+                    self._create_event_from_change_request(change_request)
+                    instance_info.state.change_requests[change_request.number] = change_request.state
+
+    def _create_event_from_change_request(self, change_request):
+        identifiers = [
```

Comment:
What are the values for `Service`, `Affected CIs` and `Impacted Services/CIs` in the jira ticket we expect them to be added here if present. Does it makes sense?

Reply:
This is still on TODO list. :(
 I was pointed to table where I can get these values, but I have to test it to see if there is all data I need...